instant-xml/src/lib.rs

```diff
@@ -10,15 +10,165 @@ pub use macros::{FromXml, ToXml};
 pub mod parse;
 
 pub trait ToXml {
-    fn write_xml<W: fmt::Write>(&self, write: &mut W) -> Result<(), Error>;
-
     fn to_xml(&self) -> Result<String, Error> {
-        let mut out = String::new();
-        self.write_xml(&mut out)?;
-        Ok(out)
+        let mut parent_prefixes = BTreeSet::new();
+        let mut output = String::new();
+        let mut serializer = Serializer {
+            parent_prefixes: &mut parent_prefixes,
+            output: &mut output,
+        };
+        self.serialize(&mut serializer, None)?;
+        Ok(output)
+    }
+
+    fn serialize(
+        &self,
+        serializer: &mut Serializer,
+        field_data: Option<&mut FieldData>,
+    ) -> Result<(), Error>;
+}
+
+macro_rules! to_xml_for_number {
+    ($typ:ty) => {
+        impl ToXml for $typ {
+            fn serialize(
+                &self,
+                serializer: &mut Serializer,
+                field_data: Option<&mut FieldData>,
+            ) -> Result<(), Error> {
+                let field_data = match field_data {
+                    Some(field_data) => field_data,
+                    None => &FieldData {
+                        field_name: stringify!($typ),
+                        field_attribute: None,
+                    },
+                };
+
+                serializer.add_open_tag(field_data);
+                serializer.output.push_str(&self.to_string());
```

Comment:
Should use `write!()` or something to write into the output instead of allocating a `String`, then writing it into the output.